tests/hfmagpsu.py

```diff
@@ -0,0 +1,122 @@
+        self.ca.set_pv_value("HFMAGPSU_01:HEATERSTATUS:SP", hs)
+
+    def _write_heater_value(self, hv):
+        self._lewis.backdoor_set_on_device("heaterValue", hv)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS HEATER OUTPUT: [" + str(hv) + "]")
+        self.ca.set_pv_value("HFMAGPSU_01:HEATERVALUE:SP", hv)
+
+    def _write_max_target(self, max):
+        self._lewis.backdoor_set_on_device("maxTarget", max)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS MAX SETTING: [" + str(max) + "]")
+        self.ca.set_pv_value("HFMAGPSU_01:MAX:SP", max)
+
+    def _write_mid_target(self, mid):
+        self._lewis.backdoor_set_on_device("midTarget", mid)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS MID SETTING: [" + str(mid) + "]")
+        self.ca.set_pv_value("HFMAGPSU_01:MID:SP", mid)
+
+    def _write_ramp_rate(self, rr):
+        self._lewis.backdoor_set_on_device("rampRate", rr)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS RAMP RATE: [" + str(rr) + "]")
+        self.ca.set_pv_value("HFMAGPSU_01:RAMPRATE:SP", rr)
+
+    def _write_pause(self, p):
+        self._lewis.backdoor_set_on_device("pause", p)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS PAUSE STATUS: [" + str(p) + "]")
+        self.ca.set_pv_value("HFMAGPSU_01:PAUSE:SP", p)
+
+    def _write_limit(self, l):
+        self._lewis.backdoor_set_on_device("limit", l)
+        self.ca.set_pv_value("HFMAGPSU_01:LIMIT:SP", l)
+        self.ca.set_pv_value("HFMAGPSU_01:LOGMESSAGE", "HH:MM:SS VOLTAGE LIMIT: [" + str(l) + "]")
+
+    def test_GIVEN_midTarget_set_WHEN_read_THEN_midTarget_is_as_expected(self):
+        expected_midTarget = 2.0
+        self._write_mid_target(expected_midTarget)
+        self.ca.assert_that_pv_is("HFMAGPSU_01:MID", expected_midTarget)
+    
+    def test_GIVEN_rampRate_set_WHEN_read_THEN_rampRate_is_as_expected(self):
+        expected_rampRate = 0.25
+        self._write_ramp_rate(expected_rampRate)
+        self.ca.assert_that_pv_is("HFMAGPSU_01:RAMPRATE", expected_rampRate)
+    
+    def test_GIVEN_maxTarget_set_WHEN_read_THEN_maxTarget_is_as_expected(self):
+        expected_maxTarget = 4.0
+        self._write_max_target(expected_maxTarget)
+        self.ca.assert_that_pv_is("HFMAGPSU_01:MAX", expected_maxTarget)  
+        
+    def test_GIVEN_limit_set_WHEN_read_THEN_limit_is_as_expected(self):
+        expected_output = 35.5
+        self._write_limit(expected_output)
+        self.ca.assert_that_pv_is("HFMAGPSU_01:LIMIT", expected_output)
+
+    def test_GIVEN_heaterValue_set_WHEN_read_THEN_heaterValue_is_as_expected(self):
+        expected_heaterValue = 15.5
+        self._write_heater_value(expected_heaterValue)
+        self.ca.assert_that_pv_is("HFMAGPSU_01:HEATERVALUE", expected_heaterValue)
+        
+    def test_GIVEN_pause_set_WHEN_read_THEN_pause_is_as_expected(self):

```

Comment:
This occasionally fails for me. After a bit of digging I found that the cause is because we expect to set the pause value via the backdoor to 'ON' but we're actually passing it `1`. This occasionally works because we are setting ti both through the backdoor and through the IOC set point, which causes a race condition which occasionally allows the test to pass.

Reply:
@Tom-Willemsen found an underlying bug here - when commands begin the same (e.g. `P` for readPause in this case) but one has additional arguments (eg `P ON` for writePause), sometimes Cmdbuilder will find the P, think it's got a command match, and build that without the arguments. He helped write a fix in command_builder.py that forces CmdBuilder to match the entire command.

Reply:
I've run the tests from other emulators that use CmdBuilder to make sure it hasn't stuffed them up. So far so good